### PR TITLE
refactors: one thing in one file

### PR DESCRIPTION
**Pull request by tg44**
Problem:

#4 

Solution:

Renamed some file from UpperCamel to notUpperCamel.
Deleted some unused models+companion objects.
Wrap some case-classes to companion objects. (So they don't leak to the namespace anymore.)
Refactored the shutdown hook.
The BitsoBook object get merged to the Book object.
I replaced a collect{}.headOption to a collectFirst somewhere.

Result:

The shutdown hook will now call the shutdown functions in parallel.

Others:
 - I think the user, and userPreferences, worth to move to a separated dir, and extracted to smaller files.
 - The fixed and daily price alerts could move to a separate dir too.
 - My OCD is hurt bcs of the EmailServiceTrait, ReCaptchaService, and ExchangeService
 - I really don't like the data dir in general, but I couldn't come up with a better solution yet...
 - Otherwise nice repo! Good work!

### CODE REVIEW

File: alerts-server/app/com/alexitc/coinalerts/core/ShutdownHandler.scala [hunk 1]
```diff
@@ -44,17 +44,15 @@ class ShutdownHandler @Inject() (application: ApplicationLifecycle) {
...
         case NonFatal(ex) =>
           logger.error("Shutdown hook failed", ex)
       }
-
-      tmp.flatMap { _ => f }
     }
 
     // wait for hooks to complete
-    Await.result(result, 1.minute)
+    Await.result(Future.sequence(calledHooks), 1.minute)
```
Comment (AlexITC) [inline alerts-server/app/com/alexitc/coinalerts/core/ShutdownHandler.scala]: Nice!

File: alerts-server/app/com/alexitc/coinalerts/data/anorm/interpreters/FixedPriceAlertFilterSQLInterpreter.scala [hunk 2]
```diff
@@ -40,24 +40,28 @@ class FixedPriceAlertFilterSQLInterpreter {
...
+
+  /**
+    * [[sql]] should be a valid SQL executable by anorm, valid SQL could be:
+    * - empty string (no conditions required)
+    * - string starting with "WHERE" like "WHERE triggered_on IS NULL"
+    *
+    * The [[sql]] could require unsafe arguments which are provided in the
+    * [[params]] list.
+    */
+  class WhereCondition private[interpreters](val sql: String, val params: List[(String, String)])
```
Comment (AlexITC) [inline alerts-server/app/com/alexitc/coinalerts/data/anorm/interpreters/FixedPriceAlertFilterSQLInterpreter.scala]: I'm not convinced on this change, while I get your point, the `WhereCondition` doesn't necessarily belongs to the `FixedPriceAlertFilterSQLInterpreter` but is more like an AST-like.

File: alerts-server/app/com/alexitc/coinalerts/parsers/FixedPriceAlertFilterParser.scala [hunk 3]
```diff
@@ -58,16 +58,19 @@ class FixedPriceAlertFilterParser {
...
 
+object FixedPriceAlertFilterParser {
 
-case class Filter(key: String, value: String) {
-  override def toString: String = s"$key=$value"
-}
-object Filter {
-  def from(string: String): Option[Filter] = {
-    Option(string.split(":"))
+  case class Filter(key: String, value: String) {
```
Comment (AlexITC) [inline alerts-server/app/com/alexitc/coinalerts/parsers/FixedPriceAlertFilterParser.scala]: In a similar way to the `WhereCondition`, I'm not convinced that the `Filter` belongs to the `FixedPriceAlertFilterParser`.

File: alerts-server/app/com/alexitc/coinalerts/services/external/MailgunEmailService.scala [hunk 4]
```diff
@@ -62,12 +63,16 @@ class MailgunEmailService @Inject() (
   }
 }
 
-case class MailgunSendEmailResponse(id: String, message: String)
-object MailgunSendEmailResponse {
-  implicit val reads: Reads[MailgunSendEmailResponse] = {
-    val builder = (JsPath \ "id").read[String] and
+object MailgunEmailService{
+
+  case class MailgunSendEmailResponse(id: String, message: String)
```
Comment (AlexITC) [inline alerts-server/app/com/alexitc/coinalerts/services/external/MailgunEmailService.scala]: As the class is nested now, `SendEmailResponse` might be adequate.

**Comment by tg44** [comment]
Okay, I split the commit then...

About the others:

`the EmailServiceTrait, ReCaptchaService, and ExchangeService` we have a trait in the common dir, and implementation in the internal dir, we have a trait and implementation in the same file in the internal dir, and we have the trait and implementation in separate files in the internal dir. I just didn't feel why we handle the same concept in 3 different ways :D

`the data dir` I have no idea. I just didn't feel this ideal. The interface defined as F[_] but the async dir can wrap only a fixed F trait. This is the first thing which annoyed me, but I can't find a better way yet. (I have an idea, but its need some more playing.) 

`case-classes to objects` We don't use them elsewhere yet. So now these are totally local classes/methods. In java/C# world maybe we would hide them as private classes. So I would reverse the question why its not belongs to those companion objects? And if they not belongs to there where they belongs?

**Comment by AlexITC** [comment]
>  I just didn't feel why we handle the same concept in 3 different ways :D

It is likely that I was in a hurry to release something and I mess up the files.

> The interface defined as F[_] but the async dir can wrap only a fixed F trait

The idea is to be able separate the blocking and non-blocking layers, having this, I can reuse the blocking layer if necessary and the non-blocking layer could be customized without affecting things that doesn't concern it.

> case-classes to objects, ..., why its not belongs to those companion objects? And if they not belongs to there where they belongs?

Because is difficult to get all choices right at the first try, I have noticed that but I have been quite busy (or lazy) to not update that.

Thanks.

**Comment by tg44** [comment]
Yapp, I dont blame you or anything, this is just a matter of taste :D (And constantly learning from others style or teaching them to yours; so we will reach the ultimate best practice once.)

We usually wrap our locally used case-classes to companion objects. (I can't show you a good example, bcs those projects are not open-sourced, and my ~1 year old projects are in critical shape prior to my actual knowledge...) But for example if you have a user related controller, I would place the login case-class to its companion (with the formatters), and give down to the service only the username and password. So the login object is not exposed if its not needed. Without a "model" dir and moving every model to the scope its mostly belongs you can every time tell that its a dbo or a dto or just some business models. And this kind of organization requires a lot less file.
I'm still not convinced about what method is the "better", I just personally hates when I open the WhateverCheckerActor file, and I find about 6 classes and objects in there. I more like when there is 2 toplevel things at most (with the same name). (I would wrap all the sealed traits to objects too bcs they  belongs together, but as I said its a matter of taste maybe.)

**Comment by AlexITC** [comment]
I have tried what you mean and I haven't continued using that approach anymore mainly because I need to change too many method signatures while adding or removing fields, as you said, it is a matter of taste, it is an interesting discussion.

The current approach models the project in layers where the models are the bottom ones (all layers could use them), the models we talk about are mostly data transfer objects.

In fact, I have tried to avoid some design mistakes on [this](https://github.com/X9Developers/block-explorer) (while they share a very similar style and it has its own issues too).

**Comment by AlexITC** [comment]
In fact, it is worse than what you said, if you look to the [UserService](https://github.com/AlexITC/crypto-coin-alerts/blob/develop/alerts-server/app/com/alexitc/coinalerts/services/UserService.scala), the `create` method receives a DTO while the `loginByEmail` and `setPreferences` methods receive the plain arguments.

**Comment by tg44** [comment]
Okay! I splitted the commits. The 2 modifications that u didn't like is not part of this.

File: alerts-server/app/com/alexitc/coinalerts/models/UserVerificationToken.scala [hunk 5]
```diff
@@ -13,12 +11,3 @@ object UserVerificationToken {
     UserVerificationToken(token)
   }
 }
-
-case class SendVerificationTokenModel(email: UserEmail)
```
Comment (AlexITC) [inline alerts-server/app/com/alexitc/coinalerts/models/UserVerificationToken.scala]: This was supposed to be used for re-sending email verification tokens but for some reason, I never completed the implementation, thanks.

**Comment by AlexITC** [comment]
I couldn't have done it better, thanks!!!

**Pull request merged**